File: paname/stats.py
"""Calculs de statistiques et de progression pour le profil utilisateur."""
from datetime import date

from .extensions import db
from .models import Exposition, Visit


def total_expos():
    return Exposition.query.count()


def user_done_count(user):
    return Visit.query.filter_by(user_id=user.id).count()


def progress(user):
    total = total_expos()
    done = user_done_count(user)
    pct = round(done / total * 100) if total else 0
    return {"total": total, "done": done, "remaining": max(total - done, 0), "pct": pct}


def fun_stats(user):
    """Statistiques 'fun' à afficher sur le profil."""
    today = date.today()
    days_left = user.days_until_26
    p = progress(user)
    remaining = p["remaining"]

    stats = {
        "days_until_26": days_left,
        "is_over_26": days_left < 0,
        "weeks_until_26": max(days_left // 7, 0),
        "expos_per_week": None,
        "expos_per_month": None,
    }

    weeks = days_left / 7 if days_left > 0 else 0
    if weeks > 0 and remaining > 0:
        stats["expos_per_week"] = round(remaining / weeks, 1)
        stats["expos_per_month"] = round(remaining / (days_left / 30), 1)

    return stats

File: src/services/stats.js
"use strict";

const { Exposition, Visit } = require("../models");

async function totalExpos() {
  return Exposition.count();
}

async function userDoneCount(user) {
  return Visit.count({ where: { user_id: user.id } });
}

async function progress(user) {
  const total = await totalExpos();
  const done = await userDoneCount(user);
  const pct = total ? Math.round((done / total) * 100) : 0;
  return { total, done, remaining: Math.max(total - done, 0), pct };
}

async function funStats(user) {
  const daysLeft = user.days_until_26;
  const p = await progress(user);
  const remaining = p.remaining;

  const stats = {
    days_until_26: daysLeft,
    is_over_26: daysLeft < 0,
    weeks_until_26: Math.max(Math.trunc(daysLeft / 7), 0),
    expos_per_week: null,
    expos_per_month: null,
  };

  const weeks = daysLeft > 0 ? daysLeft / 7 : 0;
  if (weeks > 0 && remaining > 0) {
    stats.expos_per_week = Math.round((remaining / weeks) * 10) / 10;
    stats.expos_per_month = Math.round((remaining / (daysLeft / 30)) * 10) / 10;
  }
  return stats;
}

module.exports = { totalExpos, userDoneCount, progress, funStats };
